x-pack/plugins/ml/public/application/explorer/anomalies_map.tsx

```diff
@@ -0,0 +1,246 @@
+          },
+        },
+        lineColor: {
+          type: STYLE_TYPE.DYNAMIC,
+          options: { color: '#3d3d3d', fieldMetaOptions: { isEnabled: true } },
+        },
+        lineWidth: { type: STYLE_TYPE.STATIC, options: { size: 1 } },
+      },
+      isTimeAware: true,
+    },
+    visible,
+    type: 'VECTOR',
+  };
+};
+
+interface Props {
+  anomalies: AnomaliesTableRecord[];
+  jobIds: string[];
+}
+
+interface MLEMSTermJoinConfig extends EMSTermJoinConfig {
+  jobId: string;
+}
+
+export const AnomaliesMap: FC<Props> = ({ anomalies, jobIds }) => {
+  const [layerList, setLayerList] = useState<VectorLayerDescriptor[]>([]);
+  const [EMSSuggestions, setEMSSuggestions] = useState<
+    Array<MLEMSTermJoinConfig | null> | undefined
+  >();
+  const {
+    services: { maps: mapsPlugin },
+  } = useMlKibana();
+
+  const getEMSTermSuggestions = useCallback(async (): Promise<void> => {
+    if (mapsPlugin) {
+      const suggestions = await Promise.all(
+        jobIds.map(async (jobId) => {
+          const entityValues = new Set<string>();
+          let entityName;
+          for (let i = 0; i < anomalies.length; i++) {
+            if (jobId === anomalies[i].jobId && anomalies[i].entityValue !== '') {
+              entityValues.add(anomalies[i].entityValue);
+
+              if (!entityName) {
+                entityName = anomalies[i].entityName;
+              }
+            }
+            if (
+              // convert to set so it's unique values
+              entityValues.size === MAX_ENTITY_VALUES ||
+              (anomalies.length > MAX_ENTITY_VALUES && i > Math.floor(anomalies.length / 3))
+            )
+              break;
+          }
+
+          const suggestion: EMSTermJoinConfig | null = await mapsPlugin.suggestEMSTermJoinConfig({
+            emsLayerIds: COMMON_EMS_LAYER_IDS,
+            sampleValues: Array.from(entityValues),
+            sampleValuesColumnName: entityName || '',
+          });
+          if (suggestion) {
+            return { jobId, ...suggestion };
+          }
+          return suggestion;
+        })
+      );
+
+      setEMSSuggestions(suggestions.filter((s) => s !== null));
+    }
+  }, [...jobIds]);
+
+  useEffect(
+    function getInitialEMSTermSuggestions() {
+      if (anomalies && anomalies.length > 0) {
+        getEMSTermSuggestions();
+      }
+    },
+    [...jobIds]
+  );
+
+  useEffect(
+    function setupMapLayers() {
+      // Loop through suggestions list and make a layer for each
+      if (EMSSuggestions?.length) {
+        let count = 0;
+        const layers = EMSSuggestions.reduce(function (result, suggestion) {
+          if (suggestion) {
+            const visible = count === 0;
+            result.push(getChoroplethAnomaliesLayer(anomalies, suggestion, visible));
+            count++;
+          }
+          return result;
+        }, [] as VectorLayerDescriptor[]);
+
+        setLayerList(layers);
+      }
+    },
+    [EMSSuggestions, anomalies.length]
+  );
```

Comment:
it's a sync update, hence instead of state and effect you can use memo
```suggestion
const layerList = useMemo(() => {
  if (!EMSSuggestions?.length) return;
  
  return EMSSuggestions.filter(isDefined).map((suggestion, index) => {
    return getChoroplethAnomaliesLayer(anomalies, suggestion, i > 0);
  }, [] as VectorLayerDescriptor[])
  
}, [EMSSuggestions, anomalies.length])

Reply:
Good suggestion - thanks for taking a look. Updated in d92e417a9fe57db1172fb65ad003fb6742be1ff6